src/preprocessing/passes/non_clausal_simp.cpp

```diff
@@ -0,0 +1,463 @@
+/*********************                                                        */
+/*! \file non_clausal_simp.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Aina Niemetz
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2018 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Non-clausal simplification preprocessing pass.
+ **
+ ** Run the nonclausal solver and try to solve all assigned theory literals.
+ **/
+
+#include "preprocessing/passes/non_clausal_simp.h"
+
+#include "context/cdo.h"
+#include "options/proof_options.h"
+#include "smt/smt_statistics_registry.h"
+#include "theory/theory_model.h"
+
+#include <vector>
+
+using namespace CVC4;
+using namespace CVC4::theory;
+
+namespace CVC4 {
+namespace preprocessing {
+namespace passes {
+
+/* -------------------------------------------------------------------------- */
+
+NonClausalSimp::Statistics::Statistics()
+    : d_numConstantProps(
+          "preprocessing::passes::NonClausalSimp::NumConstantProps", 0)
+{
+  smtStatisticsRegistry()->registerStat(&d_numConstantProps);
+}
+
+NonClausalSimp::Statistics::~Statistics()
+{
+  smtStatisticsRegistry()->unregisterStat(&d_numConstantProps);
+}
+
+/* -------------------------------------------------------------------------- */
+
+NonClausalSimp::NonClausalSimp(PreprocessingPassContext* preprocContext)
+    : PreprocessingPass(preprocContext, "non-clausal-simp")
+{
+}
+
+PreprocessingPassResult NonClausalSimp::applyInternal(
+    AssertionPipeline* assertionsToPreprocess)
+{
+  d_preprocContext->spendResource(options::preprocessStep());
+
+  theory::booleans::CircuitPropagator* propagator =
+      d_preprocContext->getCircuitPropagator();
+
+  if (options::unsatCores() || options::fewerPreprocessingHoles())
+  {
+    return PreprocessingPassResult::NO_CONFLICT;
```

Comment:
Hm, this is actually a good question: How should we handle such preconditions for PP passes?
The issue here is that now you have to guard each call of the PP pass, and assert it as a precondition in the PP pass. Wouldn't it be actually cleaner to keep this check here? I can move it to the top of applyInternal, though.

Reply:
I think that in the future, we don't even want to create an instance of the `NonClausalSimp` pass if `options::unsatCores() || options::fewerPreprocessingHoles()`. To avoid duplicating checks, we could have a static method in each preprocessing pass that checks for incompatible options, so that when we create the list of preprocessing passes to execute, we can leave out the incompatible ones. In the pass itself, we can just assert that the incompatible options have not been set (this can even be done in the base class). Does that make sense?

Reply:
Yes. I'll move the check to smt engine, this is in scope for this PP.

Reply:
That's a good idea. It will be more visible there.